### PR TITLE
drivers: support of 64-bit addresses from Device Tree

### PR DESCRIPTION
**Pull request by amr-sc**
Usage of 64-bit address constants from Device Tree without a UINT64_C wrapping macro results in the following warning and the cut-off of the address value:

"warning: integer constant is so large that it is unsigned"

This change adds few wrappers over the address constants coming from Device Tree to support such periphery as PLIC, MTIMER and UART for 64-bit RISC-V platforms from Syntacore

### CODE REVIEW

**Comment by amr-sc** [comment]
@nashif , @andyross , @carlescufi , @galak , @MaureenHelm , @kgugala , @pgielda , @dcpleung , folks kind reminder for a review. There are just a few lines. Thanks!

**Comment by carlocaione** [comment]
uhm, this is going to add a lot of boileplate code. @mbolivar do you think we should tweak the `DT_*` macros instead to seamlessly support these cases?

**Comment by povergoing** [comment]
Could you please show the build command so I can reproduce the warning? I am curious about how it happens because this warning won't appear in most cases?

**Comment by povergoing** [comment]
> Hi @povergoing , please apply the following patch to dts for qemu_riscv64 platform:
> 

Thanks, now I am clear about the issue. And I agreed with Carlo, this might be better handled by `DT_*` macros, otherwise we would have too many similar codes in the future.

**Review by carlocaione — CHANGES_REQUESTED** [review]
This must be fixed at `DT_*` macros level, otherwise we have to carry this fix for every 64-bit platform.

**Comment by amr-sc** [comment]
Hi @carlocaione , @povergoing , agree with you guys. 
So the right way would be to modify the include/zephyr/devicetree.h right?

**Comment by carlocaione** [comment]
> So the right way would be to modify the include/zephyr/devicetree.h right?

In theory yes, but I expect that to be a bit more complicated than that. Also @mbolivar what do you think?

**Comment by amr-sc** [comment]
Hi @mbolivar-nordic, @carlocaione ,
Fixing it on DT_ level seems very unsafe. I fall into configuration testing hell and even if I fix it to the level to satisfy CI, there are still places where it may pop up quite unexpectedly.  I just can't determine all the platforms affected. 
The problem is that the same macros are used within linker scripts and ld doesn't like those ULL. 
Sometimes in case some wrappers are used in C code around DT_ macro it may pop up even there.

I will still fix the CI just in case someone will come up with better ideas for a generic solution. Thanks!

**Comment by mbolivar-nordic** [comment]
> The problem is that the same macros are used within linker scripts and ld doesn't like those ULL.

Yep, this is why I'm not able to help here, sorry. This situation is discussed in the issue I linked

**Comment by carlocaione** [comment]
@mbolivar-nordic ok, what about adding instead some `DT64_*` macros in parallel to the `DT_*` ones? That would prevent to carry the `ifdef(CONFIG_64BIT)` around everywhere and that should cause less havoc in CI.

File: include/zephyr/devicetree.h [hunk 1]
```diff
@@ -2159,14 +2159,29 @@
...
  * @return address of the idx-th register block
  */
+#if IS_ENABLED(CONFIG_64BIT)
 #define DT_REG_ADDR_BY_IDX(node_id, idx) \
-	DT_CAT4(node_id, _REG_IDX_, idx, _VAL_ADDRESS)
+	UINT64_C(DT_REG_ADDR_BY_IDX_RAW(node_id, idx))
+#else
+#define DT_REG_ADDR_BY_IDX(node_id, idx) \
+	DT_REG_ADDR_BY_IDX_RAW(node_id, idx)
+#endif /* IS_ENABLED(CONFIG_64BIT) */
```
Comment (gmarull) [inline include/zephyr/devicetree.h:2184]: I'm not convinced about this solution. `DT_REG_ADDR*` is also used for non-64-bit buses in 64-bit systems, e.g. I2C or SPI for CS. I think we should have _typed_ macros instead, e.g. `DT_REG_ADDR_U64`. For props, `dt-schema` has fixed-integer types: https://github.com/devicetree-org/dt-schema/blob/main/dtschema/schemas/types.yaml we should likely adopt them and offer typed versions of the DT API macros, e.g. `DT_PROP_U32`, etc.

**Review by gmarull — CHANGES_REQUESTED** [review]

File: include/zephyr/devicetree.h (hunk 1 above)
Comment (amr-sc) [reply include/zephyr/devicetree.h:2184]: So guys, what will we decide? To return to boilerplate code using DT_REG_ADDR_U64 which will be a wrapper around UINT64_C macro, right? And put it to all the places where necessary?

E.g. like this:

```
#if IS_ENABLED(CONFIG_64BIT)
#define MTIMEDIV_REG	(DT_INST_REG_ADDR_U64(0)) + 4)
#define MTIME_REG	(DT_INST_REG_ADDR_U64(0) + 8)
#define MTIMECMP_REG	(DT_INST_REG_ADDR_U64(0) + 16)
#else
#define MTIMEDIV_REG	(DT_INST_REG_ADDR(0) + 4)
#define MTIME_REG	(DT_INST_REG_ADDR(0) + 8)
#define MTIMECMP_REG	(DT_INST_REG_ADDR(0) + 16)
#endif /* CONFIG_64BIT */
```

@gmarull , @carlocaione , @evgeniy-paltsev , @mbolivar-nordic

Comment (carlocaione) [reply include/zephyr/devicetree.h:2184]: not ideal but this looks reasonable to me.

Comment (amr-sc) [reply include/zephyr/devicetree.h:2184]: > Just for note - I'd say that `CONFIG_64BIT` is not related to peripheral addresses.
> 
> We may have 32bit peripheral bus (or peripheral which require 32bit / 16bit / 8bit access) on 64bit CPU and we can have 64bit peripheral bus with 32bit CPU.

Agree. Moreover, adding ULL to the address constant which is in fact limited to 32-/16-/8-bit seems to be safe, so basically we can use _U64 version unconditionally everywhere where sometimes the real 64-bit address may be passed from device tree.  And the compiler will complain in case of overflow, as before. 

So basically returning to my example above we may just stick to:
```
#define MTIMEDIV_REG	(DT_INST_REG_ADDR_U64(0) + 4)
#define MTIME_REG	(DT_INST_REG_ADDR_U64(0) + 8)
#define MTIMECMP_REG	(DT_INST_REG_ADDR_U64(0) + 16)
```

Right?

Comment (mbolivar-nordic) [reply include/zephyr/devicetree.h:2184]: >  I think we should have typed macros instead, e.g. DT_REG_ADDR_U64

This seems reasonable to me -- @galak?

**Review by carlocaione — DISMISSED** [review]

File: include/zephyr/sys/device_mmio.h [hunk 2]
```diff
@@ -116,12 +116,12 @@ struct z_device_mmio_rom {
 
 #define Z_DEVICE_MMIO_ROM_INITIALIZER(node_id) \
 	{ \
-		.addr = DT_REG_ADDR(node_id) \
+		.addr = DT_REG_ADDR_U64(node_id) \
```
Comment (gmarull) [inline include/zephyr/sys/device_mmio.h:119]: is this always 64-bit?

File: include/zephyr/devicetree.h [hunk 3]
```diff
@@ -3676,6 +3704,14 @@
  */
 #define DT_INST_REG_ADDR(inst) DT_INST_REG_ADDR_BY_IDX(inst, 0)
 
+/**
+ * @brief 64-bit version of a DT_INST_REG_ADDR macro.
+ *        Use it in cases when 64-bit address value may appear in Device Tree
+ * @param inst instance number
+ * @return instance's register block address
+ */
+#define DT_INST_REG_ADDR_U64(inst) UINT64_C(DT_INST_REG_ADDR(inst))
```
Comment (gmarull) [inline include/zephyr/devicetree.h:3713]: this is missing tests, also, `#include <stdint.h>` is missing

File: drivers/timer/riscv_machine_timer.c [hunk 4]
```diff
@@ -67,9 +67,9 @@
 #define DT_DRV_COMPAT scr_machine_timer
 #define MTIMER_HAS_DIVIDER
 
-#define MTIMEDIV_REG	(DT_INST_REG_ADDR(0) + 4)
-#define MTIME_REG	(DT_INST_REG_ADDR(0) + 8)
-#define MTIMECMP_REG	(DT_INST_REG_ADDR(0) + 16)
+#define MTIMEDIV_REG	(DT_INST_REG_ADDR_U64(0) + 4)
+#define MTIME_REG	(DT_INST_REG_ADDR_U64(0) + 8)
+#define MTIMECMP_REG	(DT_INST_REG_ADDR_U64(0) + 16)
```
Comment (gmarull) [inline drivers/timer/riscv_machine_timer.c:72]: is this always 64-bit?

**Review by gmarull — CHANGES_REQUESTED** [review]

File: include/zephyr/devicetree.h (hunk 3 above)
Comment (amr-sc) [reply include/zephyr/devicetree.h:3713]: I've made tests using the platform which is not currently upstreamed (we are paving the road for this).
Is it ok, or would you suggest some specific tests here?

Comment (amr-sc) [reply include/zephyr/devicetree.h:3713]: As for stdint, I believe it is getting included through #include <zephyr/sys/util.h> -> #include <zephyr/types.h>

Comment (mbolivar-nordic) [reply include/zephyr/devicetree.h:3713]: All macros in zephyr/devicetree.h (and its sub-headers) require test cases in tests/lib/devicetree. Usually tests/lib/devicetree/api/src/main.c.

File: include/zephyr/devicetree.h [hunk 5]
```diff
@@ -2203,6 +2203,14 @@
  */
 #define DT_REG_ADDR(node_id) DT_REG_ADDR_BY_IDX(node_id, 0)
 
+/**
+ * @brief 64-bit version of a DT_REG_ADDR macro.
+ *        Use it in cases when 64-bit address value may appear in Device Tree
```
Comment (mbolivar-nordic) [inline include/zephyr/devicetree.h:2208]: "devicetree" is preferred to "device tree"; please use this instead.

The `@brief` block should generally be one line; please separate this second line and do not indent it.

File: include/zephyr/devicetree.h [hunk 6]
```diff
@@ -2203,6 +2203,14 @@
...
+ * @param node_id node identifier
+ * @return node's register block address
```
Comment (mbolivar-nordic) [inline include/zephyr/devicetree.h:2210]: Please specify more clearly what the difference is between this value and the usual DT_REG_ADDR one.

File: include/zephyr/devicetree.h [hunk 7]
```diff
@@ -2203,6 +2203,14 @@
...
+ */
+#define DT_REG_ADDR_U64(node_id) UINT64_C(DT_REG_ADDR(node_id))
```
Comment (mbolivar-nordic) [inline include/zephyr/devicetree.h:2212]: Please split the devicetree API changes into their own commit(s), separate from the commit modifying your platform.

File: include/zephyr/devicetree.h [hunk 8]
```diff
@@ -2221,6 +2229,16 @@
 #define DT_REG_ADDR_BY_NAME(node_id, name) \
 	DT_CAT4(node_id, _REG_NAME_, name, _VAL_ADDRESS)
 
+/**
```
Comment (mbolivar-nordic) [inline include/zephyr/devicetree.h:2232]: Same comments apply here and below as to the previous macro.

**Review by mbolivar-nordic — DISMISSED** [review]

File: include/zephyr/devicetree.h (hunk 3 above)
Comment (gmarull) [reply include/zephyr/devicetree.h:3713]: > As for stdint, I believe it is getting included through #include <zephyr/sys/util.h> -> #include <zephyr/types.h>

Relying on indirect includes is a fragile approach.

File: include/zephyr/devicetree.h (hunk 5 above)
Comment (amr-sc) [reply include/zephyr/devicetree.h:2208]: Got it, thanks!

File: include/zephyr/devicetree.h (hunk 7 above)
Comment (amr-sc) [reply include/zephyr/devicetree.h:2212]: Ok, makes sense

File: include/zephyr/devicetree.h (hunk 3 above)
Comment (mbolivar-nordic) [reply include/zephyr/devicetree.h:3713]: Agreed, please include what you use (and remember to be careful to allow this header to be included from linker scripts of course).

File: include/zephyr/devicetree.h (hunk 8 above)
Comment (amr-sc) [reply include/zephyr/devicetree.h:2232]: Ok

File: include/zephyr/devicetree.h (hunk 3 above)
Comment (amr-sc) [reply include/zephyr/devicetree.h:3713]: > Relying on indirect includes is a fragile approach.
Agree, will add it thanks!

File: include/zephyr/sys/device_mmio.h (hunk 2 above)
Comment (gmarull) [reply include/zephyr/sys/device_mmio.h:119]: `addr` is of type `mm_reg_t` an alias for `uintptr_t`. This is not always unsigned 64-bit.

Comment (amr-sc) [reply include/zephyr/sys/device_mmio.h:119]: Agree, but this shouldn't be a problem, or maybe I'm missing something?

I made this change with the following assumption:
if we assign some value to 32-bit field we expect this value to fit to 32-bits, otherwise the overflow will be detected by the compiler, and adding ULL to that integer literal will not change this behavior

File: include/zephyr/devicetree.h [hunk 9]
```diff
@@ -2237,6 +2268,22 @@
...
+ * @brief 64-bit version of a DT_REG_ADDR_BY_NAME macro
+ *
+ * Original DT_REG_ADDR_BY_NAME macro will result in warnings with 64-bit
+ * addresses from devicetree, so this macro should be used
+ * in such cases instead. It's also safe to use it with 32-bit addresses.
```
Comment (carlocaione) [inline include/zephyr/devicetree.h:2276]: but if we specify that `it's also safe to use it with 32-bit addresses` the user can wonder why we need the 32-bit version at all. Can you specify here also the limitation of this 64-bit version? And possibly be more pedantic when to use one or the other version? Otherwise this is going to confuse a lot of people.

**Review by carlocaione — CHANGES_REQUESTED** [review]

File: include/zephyr/devicetree.h [hunk 10]
```diff
@@ -17,9 +17,26 @@
...
+ * be done in the preprocessor (casts aren't allowed).  Defer to the
+ * GCC internal functions where they're available.
+ *
+ * native_posix platform uses it's own version of stdint.h, where
+ * this might not be handled, so we explicitly redefine UINT64_C
+ * macro here
+ */
+#undef UINT64_C
+#define UINT64_C(_v) __UINT64_C(_v)
+#endif
```
Comment (gmarull) [inline include/zephyr/devicetree.h:38]: this is starting to become quite hacky

Comment (amr-sc) [reply include/zephyr/devicetree.h:38]: I also don't like this. 
This problem appeared when I tried to write a test on device tree, where the native_posix platform is used. In native (not Zephyr's) version of stdint.h the UINT64_C macro is defined the following way:

```
define UINT64_C(c)   c ## ULL
```

While in Zephyr we have in lib/libc/minimal/include/stdint.h:

```
#ifdef __GNUC__
/* These macros must produce constant integer expressions, which can't
 * be done in the preprocessor (casts aren't allowed).  Defer to the
 * GCC internal functions where they're available.
 */
#define INT8_C(_v) __INT8_C(_v)
#define INT16_C(_v) __INT16_C(_v)
#define INT32_C(_v) __INT32_C(_v)
#define INT64_C(_v) __INT64_C(_v)
#define INTMAX_C(_v) __INTMAX_C(_v)

#define UINT8_C(_v) __UINT8_C(_v)
#define UINT16_C(_v) __UINT16_C(_v)
#define UINT32_C(_v) __UINT32_C(_v)
#define UINT64_C(_v) __UINT64_C(_v)
#define UINTMAX_C(_v) __UINTMAX_C(_v)
#endif /* __GNUC__ */
```

So what options do we have here? Can you suggest something?

File: include/zephyr/devicetree.h (hunk 9 above)
Comment (amr-sc) [reply include/zephyr/devicetree.h:2276]: Hi @carlocaione! You're right. 
We basically can use _U64 versions instead of the original ones. 
The only reason why we don't replace the original macro is that we can't use _U64 versions inside linker scripts since ld complains on "ULL"

I can rephrase it like this:

```
 * Original AAA macro will result in warnings with 64-bit
 * addresses from devicetree, so this macro should be used
 * in such cases instead. 
 * Generally this macro could be used interchangeably with 
 * AAA macro with one exception: it couldn't be used in linker scripts.
 * The general rule of thumb is to use 64-bit version when the
 * appearance of 64-bit address is possible and when it's not 
 * conflicting with linker scripts
```

What do you say?

Comment (carlocaione) [reply include/zephyr/devicetree.h:2276]: so, IIUC the only difference is that the `_U64` version is adding the `ULL` suffix? In that case this must be explicit in the description.

Comment (amr-sc) [reply include/zephyr/devicetree.h:2276]: Ok, thanks. How about this:

```
 * This macro adds ULL postfix to the address value returned
 * by AAA macro. 
 * Generally AAA_U64 macro could be used interchangeably with 
 * AAA macro with one exception: it couldn't be used in linker scripts.
 * The general rule of thumb is to use 64-bit version when the
 * appearance of 64-bit address is possible and when it's not 
 * conflicting with linker scripts
```

Comment (carlocaione) [reply include/zephyr/devicetree.h:2276]: yes, this is better, thanks.

File: include/zephyr/devicetree.h (hunk 10 above)
Comment (amr-sc) [reply include/zephyr/devicetree.h:38]: @mbolivar-nordic , @carlocaione , what do you guys think?

Comment (amr-sc) [reply include/zephyr/devicetree.h:38]: @carlocaione , @gmarull , @mbolivar-nordic , guys, do you have any ideas? What else could we do here?

Comment (gmarull) [reply include/zephyr/devicetree.h:38]: Since DT API is used from Linker, you could have something like:

```c
/* include if !_LINKER */
#ifndef _LINKER
#include <stdint.h>
#endif

/* typed macro, document it is equivalent to DT_REG_ADDR in linker context */
#define DT_REG_ADDR_U64(node_id) DT_UINT64_C(DT_REG_ADDR(node_id))

/* ...at the bottom, with helpers */
#ifndef _LINKER
#define DT_UINT64_C(_v) UINT64_C(_v)
#else
#define DT_UINT64_C(_v)
#endif
```

(or add

Comment (amr-sc) [reply include/zephyr/devicetree.h:38]: Hi @gmarull , 
Hmmm, that's probably could solve our issue with generic solution. We probably may have add such DT_U64_C for original macros and avoid the usage of specific _U64 versions altogether. 

What do you think?

Comment (gmarull) [reply include/zephyr/devicetree.h:38]: > Hi @gmarull , Hmmm, that's probably could solve our issue with generic solution. We probably may have add such DT_U64_C for original macros and avoid the usage of specific _U64 versions altogether.
> 
> What do you think?

No, because register addresses are not always 64-bit, see e.g. I2C bus addresses.

Comment (amr-sc) [reply include/zephyr/devicetree.h:38]: But adding ULL to, say, 8-bit integer literal value will not hurt, and the user remains unaware of the form he receives the address value as far as it is a correct value. From the user's point of view he will use the same
familiar DT_REG_ADDR() macro without caring that under the hood it modifies the address constant in a harmless way.

P.S.: I'm not insisting on this approach, I'm just trying to ensure that we are on the same page. If you want to have explicit _U64 versions of DT_ macros - that's OK, let's have it, even it is slightly ambiguous in relation to 32-bit addresses (in the same way you pointed out with your I2C bus example)

Comment (amr-sc) [reply include/zephyr/devicetree.h:38]: Hi @gmarull , @carlocaione , @mbolivar-nordic , 
Guys, let me kindly summarize the remaining questions:

1. What should we do with the redefinition of UINT64_C for native_posix platform in device_tree.h? Shouldn't the native_posix platform use the same stdint.h from lib/libc/minimal/include/stdint.h as the rest of the Zephyr platforms? Currently it looks somehow ugly:

```
#ifdef __GNUC__
/*
 * This macro must produce constant integer expressions, which can't
 * be done in the preprocessor (casts aren't allowed).  Defer to the
 * GCC internal functions where they're available.
 *
 * native_posix platform uses it's own version of stdint.h, where
 * this might not be handled, so we explicitly redefine UINT64_C
 * macro here
 */
#undef UINT64_C
#define UINT64_C(_v) __UINT64_C(_v)
#endif
```

2. @gmarull suggested a nice way to detect linker context, which potentially allows us to have a universal solution for address getters instead of _U64 variants:

```
/* include if !_LINKER */
#ifndef _LINKER
#include <stdint.h>
#endif

#define DT_REG_ADDR(node_id) DT_U64_C(DT_REG_ADDR(node_id))

/* ...at the bottom, with helpers */
#ifndef _LINKER
#define DT_U64_C(_v) UINT64_C(_v)
#else
#define DT_U64_C(_v)
#endif
```

Should we try this way, or should we stay with distinct explicit _U64 variants?
Thanks a lot!

Comment (carlocaione) [reply include/zephyr/devicetree.h:38]: go for @gmarull solution, we can comment on the actual implementation.

Comment (amr-sc) [reply include/zephyr/devicetree.h:38]: Hi Guys, unfortunately, it looks like the universal solution will not work, since we have macros like:

```
#define DT_PHANDLE_BY_IDX(node_id, prop, idx) \
	DT_CAT6(node_id, _P_, prop, _IDX_, idx, _PH)
#define DT_GPIO_CTLR_BY_IDX(node_id, gpio_pha, idx) \
	DT_PHANDLE_BY_IDX(node_id, gpio_pha, idx)
#define DT_GPIO_CTLR_BY_IDX(DT_BUS(spi_dev), cs_gpios, DT_REG_ADDR(spi_dev))
```
You may see that the value of DT_REG_ADDR(spi_dev) which potentially has ULL at the end will be concatenated with "_PH" which will give some awkward names like DT_N_S_test_S_spi_33334444_P_cs_gpios_IDX_0ULL_PH_ORD

So let's stay with distinct _U64 vesrsions

Comment (amr-sc) [reply include/zephyr/devicetree.h:38]: @carlocaione , @gmarull , @mbolivar-nordic , guys please review

**Review by carlocaione — COMMENTED** [review]
LGTM but this should really be acked by @mbolivar-nordic or @gmarull

File: include/zephyr/devicetree.h [hunk 11]
```diff
@@ -17,9 +17,35 @@
...
```
Comment (gmarull) [inline include/zephyr/devicetree.h:38]: why is this required?

Comment (amr-sc) [reply include/zephyr/devicetree.h]: Hmmm.. That's very strange, but previously I've got a lot of errors for native_posx platform and associated it with different UINT64_C() macro declaration for that platform.. Now it works flawlessly. So looks like we don't need this hack anymore, thanks for your comment. Removed this

File: include/zephyr/devicetree.h [hunk 12]
```diff
@@ -17,9 +17,21 @@
...
 #include <zephyr/sys/util.h>
 
+#ifndef _ASMLANGUAGE
+#include <stdint.h>
+#endif
+
+/**
+ * @brief Macro to add ULL postfix to the devicetree address constants
+ */
+#ifndef _LINKER
```
Comment (gmarull) [inline include/zephyr/devicetree.h:29]: this should be `_LINKER` only

File: include/zephyr/devicetree.h [hunk 13]
```diff
@@ -17,9 +17,21 @@
...
+#define DT_U64_C(_v) UINT64_C(_v)
+#else
+#define DT_U64_C(_v) (_v)
+#endif
```
Comment (gmarull) [inline include/zephyr/devicetree.h:33]: please move this to the internal utilities section at the bottom of the file

File: include/zephyr/devicetree.h [hunk 14]
```diff
@@ -2219,6 +2231,22 @@
...
+ * This macro adds ULL postfix to the address value returned
+ * by DT_REG_ADDR macro.
+ * Generally DT_REG_ADDR_U64 macro could be used interchangeably with
+ * DT_REG_ADDR macro with one exception: it couldn't be used in
+ * linker scripts.
+ * The general rule of thumb is to use 64-bit version when the
+ * appearance of 64-bit address is possible and when it's not
+ * conflicting with linker scripts
+ * @param node_id node identifier
+ * @return node's register block address
```
Comment (gmarull) [inline include/zephyr/devicetree.h:2246]:
```suggestion
/**
 * @brief 64-bit version of DT_REG_ADDR()
 *
 * This macro version adds the appropriate suffix for 64-bit unsigned integer literals.
 * Note that this macro is equivalent to DT_REG_ADDR() in linker context.
 *
 * @param node_id node identifier
 * @return node's register block address
```

Description needs to be more concise, a proposal above ^

**Review by gmarull — CHANGES_REQUESTED** [review]

File: include/zephyr/devicetree.h (hunk 12 above)
Comment (amr-sc) [reply include/zephyr/devicetree.h:29]: Right, thanks!

File: include/zephyr/devicetree.h (hunk 14 above)
Comment (amr-sc) [reply include/zephyr/devicetree.h:2246]: Thanks, will change it

Comment (amr-sc) [reply include/zephyr/devicetree.h]: Hi @gmarull , I've addressed your comments

**Review by carlocaione — DISMISSED** [review]

**Comment by amr-sc** [comment]
Hi @mbolivar-nordic , would you please take a look at what we finally got here? Thanks!

File: include/zephyr/devicetree.h [hunk 15]
```diff
@@ -17,9 +17,12 @@
 #define DEVICETREE_H
 
 #include <devicetree_generated.h>
-
 #include <zephyr/sys/util.h>
 
+#ifndef _ASMLANGUAGE
+#include <stdint.h>
+#endif
+
```
Comment (gmarull) [inline include/zephyr/devicetree.h:25]:
```suggestion
#ifndef _LINKER
#include <stdint.h>
#endif

```

also, move after devicetree_generated.h

File: include/zephyr/devicetree.h [hunk 16]
```diff
@@ -4172,6 +4230,15 @@
 #define DT_INST_NODE_HAS_PROP_AND_OR(inst, prop) \
 	DT_INST_NODE_HAS_PROP(inst, prop) ||
 
+/**
+ * @brief Macro to add ULL postfix to the devicetree address constants
+ */
```
Comment (gmarull) [inline include/zephyr/devicetree.h:4235]:
```suggestion
/**
 * @def DT_U64_C
 * @brief Macro to add ULL postfix to the devicetree address constants
 */
```

because comment is not attached to the macro

File: include/zephyr/sys/device_mmio.h [hunk 17]
```diff
@@ -116,12 +116,12 @@ struct z_device_mmio_rom {
...
 	}
 
 #define Z_DEVICE_MMIO_NAMED_ROM_INITIALIZER(name, node_id) \
 	{ \
-		.addr = DT_REG_ADDR_BY_NAME(node_id, name) \
+		.addr = DT_REG_ADDR_BY_NAME_U64(node_id, name) \
```
Comment (gmarull) [inline include/zephyr/sys/device_mmio.h:124]: you should add explicit casts to `mm_reg_t`, since this is really a `uintptr_t`, not always 64-bit.

File: tests/lib/devicetree/api/app.overlay [hunk 18]
```diff
@@ -621,4 +621,16 @@
 			val = "XA XPLUS XB", "XC XPLUS XD", "XA XMINUS XB", "XC XMINUS XD";
 		};
 	};
+
+	test_64 {
+		#address-cells = < 2 >;
+		#size-cells = < 0 >;
+
+		test_reg_64: reg-holder-64@ffffffff11223344 {
+			compatible = "vnd,reg-holder-64";
```
Comment (gmarull) [inline tests/lib/devicetree/api/app.overlay:630]: this binding doesn't seem to exist

**Review by gmarull — CHANGES_REQUESTED** [review]

**Comment by gmarull** [comment]
please fix commit title to `devicetree` as well

File: include/zephyr/devicetree.h (hunk 15 above)
Comment (amr-sc) [reply include/zephyr/devicetree.h:25]: Ok, will do, thanks!

File: include/zephyr/devicetree.h [hunk 19]
```diff
@@ -17,9 +17,12 @@
 #define DEVICETREE_H
 
 #include <devicetree_generated.h>
-
```
[the remark this answers is not in the document]
Comment (amr-sc) [reply include/zephyr/devicetree.h:20]: Just removed the empty line between the includes

File: include/zephyr/devicetree.h (hunk 16 above)
Comment (amr-sc) [reply include/zephyr/devicetree.h:4235]: Thanks, improved

File: include/zephyr/sys/device_mmio.h (hunk 17 above)
Comment (amr-sc) [reply include/zephyr/sys/device_mmio.h:124]: Sure, thanks

File: tests/lib/devicetree/api/app.overlay (hunk 18 above)
Comment (amr-sc) [reply tests/lib/devicetree/api/app.overlay:630]: Oh! Indeed, thanks!

File: include/zephyr/devicetree.h (hunk 12 above)
Comment (amr-sc) [reply include/zephyr/devicetree.h]: The <stdint.h> is getting included with #ifndef _ASMLANGUAGE, e.g in zephyr/sys/util.h.
Otherwise we'll get the following error:

```
...
 /__w/zephyr/zephyr/lib/libc/minimal/include/stdint.h: Assembler messages:
/__w/zephyr/zephyr/lib/libc/minimal/include/stdint.h:72: Error: bad instruction `typedef signed char int8_t'
...
```
see https://github.com/zephyrproject-rtos/zephyr/actions/runs/5198990587/jobs/9375804140

**Comment by amr-sc** [comment]
Hi @gmarull , @mbolivar-nordic , @carlocaione 
Guys, I've addressed the comments

**Review by carlocaione — DISMISSED** [review]

**Comment by amr-sc** [comment]
@gmarull , @mbolivar-nordic , guys, kind reminder for a review. I've addressed your comments.
Thanks!

File: include/zephyr/devicetree.h [hunk 20]
```diff
@@ -17,7 +17,9 @@
...
+#ifndef _ASMLANGUAGE
+#include <stdint.h>
+#endif
```
Comment (gmarull) [inline include/zephyr/devicetree.h:22]:
```suggestion
#if !defined(_LINKER) && !defined(_ASMLANGUAGE)
#include <stdint.h>
#endif

```

File: include/zephyr/devicetree.h (hunk 19 above)
[the remark this answers is not in the document]
Comment (gmarull) [reply include/zephyr/devicetree.h:20]: this is intentional:

- local headers
- empty
- stdc headers
- empty
- zephyr headers
- empty
- 3rd party headers

File: include/zephyr/devicetree.h [hunk 21]
```diff
@@ -4172,6 +4229,16 @@
 #define DT_INST_NODE_HAS_PROP_AND_OR(inst, prop) \
 	DT_INST_NODE_HAS_PROP(inst, prop) ||
 
+/**
+ * @def DT_U64_C
+ * @brief Macro to add ULL postfix to the devicetree address constants
+ */
+#ifdef _LINKER
```
Comment (gmarull) [inline include/zephyr/devicetree.h:4236]:
```suggestion
#if defined(_LINKER) || defined(_ASMLANGUAGE)
```

I guess this needs to be like this if stdint.h can't be used in ASM contect

File: include/zephyr/devicetree.h [hunk 22]
```diff
@@ -3697,6 +3741,19 @@
  */
 #define DT_INST_REG_ADDR(inst) DT_INST_REG_ADDR_BY_IDX(inst, 0)
 
+/**
+ * @brief 64-bit version of DT_INST_REG_ADDR()
+ *
+ * This macro version adds the appropriate suffix for 64-bit unsigned
+ * integer literals.
+ * Note that this macro is equivalent to DT_INST_REG_ADDR() in
+ * linker context.
```
Comment (gmarull) [inline include/zephyr/devicetree.h:3750]:
```suggestion
 * linker/ASM context.
```
? (see my other comments)

File: include/zephyr/devicetree.h (hunk 20 above)
Comment (amr-sc) [reply include/zephyr/devicetree.h:22]: Fixed, thanks

**Review by gmarull — APPROVED** [review]

**Review by carlocaione — APPROVED** [review]

**Comment by amr-sc** [comment]
Hi @mbolivar-nordic , kind reminder for a review

**Comment by amr-sc** [comment]
Hi @mbolivar-nordic , please review. We can't merge this change without your assistance. Thanks!

**Comment by amr-sc** [comment]
Hi @mbolivar-nordic , could you please take a look at the changes you requested?
Thanks!

**Comment by amr-sc** [comment]
Hi @mbolivar-nordic , your review will be very much appreciated. Thanks in advance!

**Comment by fabiobaltieri** [comment]
@galak @mbolivar-nordic could you do another pass?

**Review by de-nordic — APPROVED** [review]

**Comment by amr-sc** [comment]
Hi @carlocaione , @fabiobaltieri , @gmarull , @de-nordic, @mbolivar-nordic  looks like everything is now ready for the merge. 
However I'm not authorized to do it myself. Could you please help with this? Thanks!

**Comment by de-nordic** [comment]
> Hi @carlocaione , @fabiobaltieri , @gmarull , @de-nordic, @mbolivar-nordic looks like everything is now ready for the merge. However I'm not authorized to do it myself. Could you please help with this? Thanks!

Now everything is in the hands of Epigishmog.

**Review by fabiobaltieri — APPROVED** [review]

**Review by povergoing — APPROVED** [review]

**Comment by nashif** [comment]
@galak

**Comment by amr-sc** [comment]
Hi @galak ,

Looks like we are ready to go. Are you authorized to merge this PR? Thanks!

**Comment by galak** [comment]
@amr-sc I'll take a look at this.

**Review by galak — APPROVED** [review]
